balancer.go

```diff
@@ -158,3 +160,98 @@ func (h *Hash) Balance(msg Message, partitions ...int) (partition int) {
 
 	return
 }
+
+type randomBalancer struct {
+	mock int // mocked return value, used for testing
+}
+
+func (b randomBalancer) Balance(msg Message, partitions ...int) (partition int) {
+	if b.mock != 0 {
+		return b.mock
+	}
+	return partitions[rand.Int()%len(partitions)]
+}
+
+// CRC32Balancer is a Balancer that uses the CRC32 hash function to determine which
+// partition to route messages to.  This ensures that messages with the same key
+// are routed to the same partition.  A random partition will be selected if either
+// the key is nil, or an the empty byte slice.
+//
+// This balancer strategy is the default partitioner used by the librdkafka library
+// and the language bindings that are built on top of it, including the
+// github.com/confluentinc/confluent-kafka-go Go package.
+type CRC32Balancer struct {
+	rb randomBalancer
+}
+
+func (b CRC32Balancer) Balance(msg Message, partitions ...int) (partition int) {
+	if len(msg.Key) == 0 {
+		return b.rb.Balance(msg, partitions...)
+	}
+
+	idx := crc32.ChecksumIEEE(msg.Key) % uint32(len(partitions))
+	return partitions[idx]
+}
+
+// Murmur2Balancer is a Balancer that uses the Murmur2 hash function to determine
+// which partition to route messages to.  This ensures that messages with the same
+// key are routed to the same partition.  A round robin strategy will be used if
+// the key is nil.
+//
+// This balancer strategy is the default partitioner used by the Java library.
+type Murmur2Balancer struct {
+	rr RoundRobin
```

Comment:
Similarly, this balancer comes in two flavors, `murmur2` and `murmur2_random`.  Might make sense to decouple the RR.

See https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md